src/natcap/invest/seasonal_water_yield/seasonal_water_yield.py

```diff
@@ -879,133 +875,178 @@ def _calculate_annual_qfi(qfm_path_list, target_qf_path):
+        #
+        #       s_i^2 / a_im * exp(0.8 * s_i / a_im) * E1(s_i / a_im)
+        #
+        # 3a. Where s_i = 0, this term should equal 0 (per conversation with
+        #     Rafa). But, evaluating in numpy you get NaN and a warning
+        #     because E1(0 / a_im) = infinity.
+        #
+        #     Solution: Preemptively set the result to 0 where s_i = 0 in
+        #     order to avoid calculations with infinity.
+        #
+        # 3b. When the ratio s_i / a_im becomes large, this term approaches 0.
+        #     [NOTE: I don't know how to prove this mathematically, but it
+        #     holds true when I tested with reasonable values of s_i and a_im].
+        #     The exp() term becomes very large, while the E1() term becomes
+        #     very small. When (s_i / a_im) > 880 ish, exp(0.8 * s_i / a_im)
+        #     will overflow and round up to infinity. Meanwhile, E1(s_i / a_im)
+        #     will become so small that it rounds down to 0.
+        #
+        #     Solution: Catch the overflow warning, and set the term to 0
+        #     anywhere that overflow happens.
+        #
+        # 3c. Otherwise, evaluate the term as usual.
+
+        # qf_im is the quickflow at pixel i on month m
+        qf_im = numpy.full(p_im.shape, TARGET_NODATA, dtype=numpy.float32)
+
+        # case 1: there is no precipitation where both p_im and n_m are
+        # defined and equal to zero.
+        case_1_mask = ~precip_mask
+        qf_im[case_1_mask] = 0
+
+        # case 2: there is precipitation where both p_im and n_m are
+        # defined and greater than zero.
+        case_2_mask = precip_mask & stream_mask
+        qf_im[case_2_mask] = p_im[case_2_mask]
+
+        # case 3
+        # only where all four inputs are defined
+        case_3_mask = valid_mask & precip_mask & ~stream_mask
+
+        term_result = numpy.full(
+            qf_im[case_3_mask].shape, TARGET_NODATA, dtype=numpy.float32)
 
         # a_im is the mean rain depth on a rainy day at pixel i on month m
         # the 25.4 converts inches to mm since Si is in inches
-        a_im = numpy.empty(valid_n_events.shape)
-        a_im = p_im[valid_mask] / (valid_n_events * 25.4)
-        qf_im = numpy.empty(p_im.shape)
-        qf_im[:] = qf_nodata
-
-        # Precompute the last two terms in quickflow so we can handle a
-        # numerical instability when s_i is large and/or a_im is small
-        # on large valid_si/a_im this number will be zero and the latter
-        # exponent will also be zero because of a divide by zero. rather than
-        # raise that numerical warning, just handle it manually
-        E1 = scipy.special.expn(1, valid_si / a_im)
-        E1[valid_si == 0] = 0
-        nonzero_e1_mask = E1 != 0
-        exp_result = numpy.zeros(valid_si.shape)
-        exp_result[nonzero_e1_mask] = numpy.exp(
-            (0.8 * valid_si[nonzero_e1_mask]) / a_im[nonzero_e1_mask] +
-            numpy.log(E1[nonzero_e1_mask]))
```

Comment:
The old implementation took advantage of exponent math to avoid overflow: By rewriting `exp(0.8 * s_i / a_im) * E1(s_i / a_im)` as the equivalent `exp(0.8 * s_i / a_im + log(E1(s_i / a_im)))`, the result of `exp` wouldn't get so large. But then, there's another edge case where E1 = 0 and log(E1) = infinity. That was handled with the `nonzero_e1_mask` above.

I replaced that with handling the overflow warning, below. It's closer to the original equation, and IMO it's clearer what's going on.